### ISSUE TITLE
[dcl.constexpr] Uninitialized non-static data members of fundamental type CWG2558

### ISSUE THREAD

**Issue by xmh0511** [cplusplus/draft#5422]
Consider this example
````cpp
struct A{
   constexpr A(){}
   int value;
};
constexpr A obj;  // #1
````
In C++17 standard, we have the restriction on the constructor that is declared with `constexpr`. [[dcl.constexpr] p4](https://timsong-cpp.github.io/cppwp/n4659/dcl.constexpr#4)
> The definition of a constexpr constructor shall satisfy the following requirements:
>> - [...]
>> - every non-variant non-static data member and base class subobject shall be initialized 

Since the member `value` does not have a default initializer or specified its initialization with the mem-initializer, thus its default initialization performs no initialization.  From this perspective, we can justify that `#1` is ill-formed. However, in the current draft, we do not have this restriction anymore, and I do not find any replaceable rules. The full-expression of the initialization does not violate [expr.const] p5 nor [expr.const] p11, however, the example is rejected by all implementations. The reason for the rejection is basically:     
````
subobject of type 'int' is not initialized
'A()' is not a constant expression because it refers to an incompletely initialized variable
````

We do not have the restrictions in either [dcl.constexpr] or [expr.const]. Is the revision of the draft omit the restriction?

**Comment by JohelEGP** [comment]
https://cplusplus.github.io/CWG/issues/2536.html seems related.

**Comment by jensmaurer** [comment]
It was intentional that we can have core constant expressions with undefined data. We might not want to have undefined results when initializing a constexpr variable, though. See the issue cross-linked in CWG2558 for details.

**Comment by frederick-vs-ja** [comment]
> If the omission will be identified to be an accident, the proposed wording might be that add an item in [expr.const] p5, which is:
> 
> > a default-initialization that performs no initialization.

No. What you suggested is partially reverting [P1331R2](https://wg21.link/p1331r2), i.e. still disallowing trivial default initialization in constant evaluation while just allowing them in `constexpr` functions (like `throw`-expressions). I don't think this is wanted.

----

I think the resolution would be like this (added to [expr.cons]/11):
> (11.?) - if the value is of scalar type, it does not contain an indeterminate value.
> (11.?) - if the value is of union type, it has one active member.

**Comment by xmh0511** [comment]
> still disallowing trivial default initialization in constant evaluation
That is what the implementations do here.  I'm not seeing the wording in P1331R2 that supports this behavior. 

If we want to augment [expr.cons]/11 to fix this issue. I think [expr.cons]/11 should first be fixed since the wording is not clear
> a prvalue core constant expression whose value satisfies the following constraints:
>> - if the value is an object of class type. 

How could the value be an object? At least, an object can have a lifetime, how could the value of a prvalue have?

**Comment by frederick-vs-ja** [comment]
> > still disallowing trivial default initialization in constant evaluation
> 
> That is what the implementations do here. I'm not seeing the wording in P1331R2 that supports this behavior.

Implementations do allow it. See [here](https://godbolt.org/z/715bY737e). In this example, `x` is trivially default-initialized.
```C++
constexpr int foo = []{
    int x;
    x = 42;
    return x;
}();
```

> How could the value be an object? At least, an object can have a lifetime, how could the value of a prvalue have?

I suspect that using "object" may be OK here, as the prvalue will eventually be materialized.

**Comment by frederick-vs-ja** [comment]
> > still disallowing trivial default initialization in constant evaluation
> 
> That is what the implementations do here. I'm not seeing the wording in P1331R2 that supports this behavior.

Implementations do allow it. See [here](https://godbolt.org/z/715bY737e). In this example, `x` is trivially default-initialized.
```C++
constexpr int foo = []{
    int x;
    x = 42;
    return x;
}();
```

> How could the value be an object? At least, an object can have a lifetime, how could the value of a prvalue have?

I suspect that using "object" may be OK here, as the prvalue will eventually be materialized.

**Comment by xmh0511** [comment]
> I suspect that using "object" may be OK here, as the prvalue will eventually be materialized. 
 A prvalue should keep its meaning that to be a  pure value until given a result object, see [class.temporary] p2.
> The materialization of a temporary object is generally delayed as long as possible in order to avoid creating unnecessary temporary objects[.](https://eel.is/c++draft/class.temporary#2.sentence-1)

If we discussed any object in a prvalue, the design of prvalue is meaningless. Again, an object occupy storage, prvalue does not. 

To clarify this point, we might say
> the object... in the result object of the prvalue.

**Comment by frederick-vs-ja** [comment]
The part for CWG2558 is resolved (a union value without an active member can be a constant expression, while an indeterminate scalar value can't). But it's still a bit weird to say a subobject of a prvalue.

> To clarify this point, we might say
> 
> > the object... in the result object of the prvalue.

I think it's safe to check the result object, since in contexts needing to determine whether a expression of a class or array type is a constant expression, the result object must exist.

It might be better to replace the occurence of "the value" with "the result object", not sure whether this is editorial.

**Comment by t3nsor** [comment]
I suppose we could make this a little bit more clear. In [expr.const]/22, where it says "a prvalue core constant expression whose value satisfies the following constraints", we could instead say something like "a prvalue core constant expression whose result object would satisfy the following constraints". But this needs to be a core issue, not an editorial issue.

While we're on this topic, I'd also like to fix [expr.const]/6.1, which says "the full-expression of its initialization is a constant expression when interpreted as a constant-expression,". The initialization of a variable doesn't have a value category, because it's not an expression. (note that this wording was left unchanged by P2686R5)

**Comment by frederick-vs-ja** [comment]
> In [expr.const]/22, where it says "a prvalue core constant expression whose value satisfies the following constraints", we could instead say something like "a prvalue core constant expression whose result object would satisfy the following constraints".

This is what I intended to say, see #6267. I'm not sure whether there's should be another CWG issue.

> While we're on this topic, I'd also like to fix [expr.const]/6.1, which says "the full-expression of its initialization is a constant expression when interpreted as a constant-expression,". The initialization of a variable doesn't have a value category, because it's not an expression. (note that this wording was left unchanged by P2686R5)

This was reported in cplusplus/CWG#460... but no formal CWG issue has been created.
